### PR TITLE
MAG240M distributed training

### PR DESCRIPTION
**Pull request by ZenoTan**
## Description

- Run _ogbn-mag240m_ dataset in Quiver.
- Optimize the performance with adaptive data allocation.
- Enable distributed feature communication.
- Prepare the evaluation results for our paper.

## Milestones

- [x] NCCL primitives with RDMA support
- [x] Feature exchange
- [x] Quiver's feature allocation algorithm
- [x] Quiver's 240M example according to PyG single-device training (DGL omitted because it is similar)
- [x] Quiver's 240M multi-GPU example (run on komodo)
- [x] Quiver's 240M multi-node example (run on komodos)
- [x] DGL and P3 multi-node baselines (simulation, will add full experiments on the cloud)
- [x] Documentation for distributed design and APIs

## Notes
- A simple algorithm to partition the feature data, and we can improve it
- To repeatedly run experiments, we store the features to cache on disks, but there is an option to generate online
- There could be a way to fully utilise NCCL bidirectional bandwidth, especially when the number of nodes is not the power of 2

### CODE REVIEW

File: benchmarks/ogbn-mag240m/preprocess.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+import os.path as osp
+from torch._C import device
+from torch_sparse import SparseTensor
+import time
+import numpy as np
+from quiver.partition import partition_with_replication, partition_without_replication, select_nodes
+
+
+def get_nonzero():
+    dataset = MAG240MDataset("/data/mag")
```
Comment (luomai) [inline benchmarks/ogbn-mag240m/preprocess.py:15]: Pass the argument instead of hardcoding it.

File: benchmarks/ogbn-mag240m/preprocess.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        cnt += 1
+        t1 = time.time()
+        print(t1 - t0)
+    torch.save(feat, processed_file)
+
+
+def preprocess(host, host_size, p2p_group, p2p_size):
+    GPU_CACHE_GB = 20
+    CPU_CACHE_GB = 40
+    dataset = MAG240MDataset("/data/mag")
```
Comment (luomai) [inline benchmarks/ogbn-mag240m/preprocess.py:72]: The same above.

File: benchmarks/ogbn-mag240m/preprocess.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        host_probs_sum[h] = probs_sum
+    gpu_size = GPU_CACHE_GB * 1024 * 1024 * 1024 // (768 * 4)
+    cpu_size = CPU_CACHE_GB * 1024 * 1024 * 1024 // (768 * 4)
+    _, nz = select_nodes(0, host_probs_sum, None)
+    res = partition_without_replication(0, host_probs_sum, nz.squeeze())
+    global2host = torch.zeros(nodes, dtype=torch.int32, device=0) - 1
+    for h in range(host_size):
+        global2host[res[h]] = h
+    torch.save(global2host.cpu(),
+               '/data/mag/mag240m_kddcup2021/processed/paper/global2host.pt')
```
Comment (luomai) [inline benchmarks/ogbn-mag240m/preprocess.py:135]: Same.

File: benchmarks/ogbn-mag240m/preprocess.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    local_p2p_probs = [prob[local_all] for prob in host_p2p_probs[host]]
+    local_res = partition_without_replication(0, local_p2p_probs,
+                                              local_gpu_order)
+    local_cpu_ids = local_all[local_cpu_order]
+    local_gpu_orders = torch.cat(local_res)
+    local_gpu_ids = [local_all[r] for r in local_res]
+    local_orders = torch.cat((local_gpu_orders, local_cpu_order))
+    torch.save(
+        local_orders.cpu(),
+        f'/data/mag/mag240m_kddcup2021/processed/paper/local_order{host}.pt')
```
Comment (luomai) [inline benchmarks/ogbn-mag240m/preprocess.py:162]: Same.

File: benchmarks/ogbn-mag240m/train_pyg.py [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+        super().__init__()
+        self.save_hyperparameters()
+        self.model = model.lower()
+        self.dropout = dropout
+
+        self.convs = ModuleList()
+        self.norms = ModuleList()
+        self.skips = ModuleList()
+
+        if self.model == 'gat':
```
Comment (luomai) [inline benchmarks/ogbn-mag240m/train_pyg.py:165]: Can we show the benefit of using Quiver for GAT?

File: srcs/cpp/src/quiver/cuda/quiver_comm.cu [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        void *ptr;
+        ptr_type(tensor, &ptr, &type);
+        ncclAllReduce(ptr, ptr, tensor.numel(), type, ncclSum, nccl_comm,
+                      stream);
+    }
+
+  private:
+    int rank;
+    int size;
+    ncclComm_t nccl_comm;
```
Comment (luomai) [inline srcs/cpp/src/quiver/cuda/quiver_comm.cu:69]: Do we need to deconstruct this communicator?

File: srcs/python/quiver/comm.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        remote_sizes = torch.zeros(self.size * self.size, dtype=torch.int64)
+        for host in range(self.table.hosts):
+            ids = host2ids[host]
+            remote_peer = self.table.remote_peer(self.rank, host)
+            if ids is not None and remote_peer != self.rank:
+                remote_sizes[self.rank * self.size + remote_peer] = ids.size(0)
+        remote_sizes = remote_sizes.to(self.device)
+        self.allreduce(remote_sizes)
+        comm_mat = self.table.get_comm_mat(remote_sizes)
+        steps = schedule(comm_mat, self.table)
```
Comment (luomai) [inline srcs/python/quiver/comm.py:149]: Can you add comments to the schedule function?

File: srcs/python/quiver/feature.py [hunk 8]
```diff
@@ -317,3 +464,112 @@ def lazy_init_from_ipc_handle(self):
...
+            host_nodes = self.global2local[host_nodes]
+            host_ids.append(host_nodes)
+            host_orders.append(host_order)
+        torch.cuda.current_stream().synchronize()
+
+        return host_ids, host_orders
+
+
+class DistFeature:
+    """DistFeature stores local features and it can fetch remote features by the network.
```
Comment (luomai) [inline srcs/python/quiver/feature.py:538]: Not clear what does this means? This is a virtual tensor?

File: tests/python/cuda/test_comm.py [hunk 9]
```diff
@@ -0,0 +1,391 @@
+import torch
+import torch.multiprocessing as mp
+import torch_quiver as torch_qv
+import quiver
+import torch.distributed as dist
+import os
+import time
+
+LOCAL_ADDR = '192.168.0.78'
```
Comment (luomai) [inline tests/python/cuda/test_comm.py:9]: localhost or 127.0.0.1?

**Comment by luomai** [comment]
Is this PR ready to merge?

**Comment by ZenoTan** [comment]
> Is this PR ready to merge?

Not yet.

**Pull request closed without merging**